### PR TITLE
feat(coil-extension): add FramesService

### PR DESCRIPTION
**Pull request by sublimator**
S̶e̶e̶m̶s̶ ̶c̶h̶r̶o̶m̶e̶ ̶e̶x̶t̶e̶n̶s̶i̶o̶n̶ ̶A̶P̶I̶ ̶h̶a̶s̶ ̶n̶o̶ ̶w̶a̶y̶ ̶o̶f̶ ̶d̶e̶t̶e̶c̶t̶i̶n̶g̶ ̶u̶r̶l̶ ̶c̶h̶a̶n̶g̶e̶ ̶e̶x̶c̶e̶p̶t̶ ̶f̶o̶r̶ ̶t̶h̶e̶ ̶t̶o̶p̶ ̶f̶r̶a̶m̶e̶.̶
(edit: see chrome.webNavigation [apis](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webNavigation))

Interestingly, in the chrome.tabs.onUpdated listener any *already loaded* child iframe readyState changes are reported. i.e. *seemingly* not the initial loading -> interactive -> complete for child frames, but subsequent ones (where say the window.location.href is set to some other url in a timeout )

It seems the title and url in the changeInfo are only for the window.top.

So say you embed a youtube iframe, then you'll have to add some other kind of url changed detection mechanism to detect if the video changes in the content script.

There seems to be various options. None particularly fantastic.
1. Polling
2. monkey patching history.pushState etc so you can add new window events.
    * This would need to be done by `<script>...</script>` injection as while the content script JS context shares underlying DOM it doesn't share the exact same JS context. One must communicate via DOM events/ window messages.
3. adding a MutationObserver to the body and detecting href changes in the mutations (sounds pretty perf destroying and ... )

### 17/1/2020

chrome.webNavigation seems to have [apis](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webNavigation) that deal with `<iframe>`s 

Note also that MessageSender interface (message sender is available in onMessage listeners) has a frameId property:
```
        /**
         * The frame that opened the connection. 0 for top-level frames, positive for child frames. This will only be set when tab is set.
         * @since Chrome 41.
         */
        frameId?: number;
    }
```
#### Youtube Notes 
`<iframe>` embeds only play for one video
If you want multiple videos you must do a "playlist" embed
In neither case does the url change.
Playlist embed url -> Channel detection would need to be added. 
**NOTE** A playlist could include videos owned by creators than the list creator
  * In that case, how could the currently playing video be detected ? 
     * This seems far from trivial

#### Impl idea 1

   A monetization tag observer will be run in the content script of *direct*
   descendants of the top level frame of each tab.

   Upon detection of a tag, the content script will send a message containing
   the tabId and frameId to the background page (implied in onMessage listener
   MessageSender attributes)

   ( Note that there is an API to get the parentFrameId for a given tabId/frameId
     which is the composite key that uniquely identifies a frame. Later if we want Russian doles ... )

   The background page will chrome.tabs.sendMessage a message to the top level
   frame for the given tabId to initiate a search for iframes. This means we
   will not need a MutationObserver on the body which would negatively effect
   performance.

   The content script of the top frame will then querySelectorAll('iframe')
   parsing the `@allow` attribute to detect "monetization"

   For each iframe found a random 128 bit id will be generated to use as a
   correlation id/allow token.

   The background page will notified of this list of {token: string, allow: boolean}

   Each iframe will then be sent the allow token via iframe.contentWindow.postMessage().

   Inside a window 'message' event handler the tab child frames will then each report
   their token/tabId/frameId to the background page via
   chrome.runtime.sendMessage from the content script 
   
   For each token/tabId/frameId message sent to the background page it's
   possible to take action on monetization streams. Stopping (rare case of
   dynamic allow attribute changing) etc

   This allow token should probably be sent along with any startWebMonetization
   requests from child frames? 
  
  To truly catch all dynamic changes of `iframe[@allow]` one could add
  MutationObserver's (with {childList: false, attributes: true, subtree: false, the point is to avoid having MutationObserver on body} options) during the `<meta>` observation driven `querySelectorAll("iframe")` detection round. Perhaps the iframes/observers could be 
  kept in a WeakMap()

##### Notes on correlationId/"allow token"

It is a way of correlating <iframe> elements in top level content scripts with their corresponding content scripts. How can we get a given tabId and frameId (as per webNavigation apis and in the MessageSender arg of runtime.onMessage event handlers) for a certain iframe element? 

We can use frame.contentWindow.postMessage to send it a random uuid4 (already sent to the background page). The content script can monitor for these messages.

When an iframe wants to start monetization it needs to send this random uuid as a kind of correlationId and "allow token" (128 bits essentially unguessable and these are only given to iframes with monetization in their @allow feature policy )

So it serves as a correlationId and given we are using window.postMessage (only way I could see), we want these to be not guessable.

### CODE REVIEW

File: packages/coil-extension/manifest.json [hunk 1]
```diff
@@ -31,5 +31,5 @@
       "run_at": "document_start"
     }
   ],
-  "permissions": [ "activeTab", "storage", "notifications", "tabs", "<all_urls>" ]
+  "permissions": [ "webNavigation", "activeTab", "storage", "notifications", "tabs", "<all_urls>" ]
```
[the remark this answers is not in the document]
Comment (sublimator) [reply packages/coil-extension/manifest.json]: I suppose it will unfortunately ?

[the remark this answers is not in the document]
Comment (sublimator) [reply packages/coil-extension/manifest.json]: There's not really any other way that will work. There's no official API for tracking url changes  we could use in the content script (disclaimer: that I've seen thus far, but I had a pretty decent look)

File: packages/coil-extension/src/content/services/Frames.ts [hunk 2]
```diff
@@ -1,21 +1,106 @@
...
+    if (this.isTopFrame) {
+      setTimeout(() => {
+        // Can't really use *= or ~= because of potential monetization in a string etc
+        // need an allow syntax parser ....
+        const frames = document.querySelectorAll<HTMLIFrameElement>(
+          'iframe[allow]'
+        )
+        Array.from(frames).forEach((frame, ix) => {
+          // WOULD need to parse here ...
+          if (!frame.allow.match(/.*monetization.*/)) {
```
[the remark this answers is not in the document]
Comment (sublimator) [reply packages/coil-extension/src/content/services/Frames.ts]: > WOULD need to parse here ...
Ja, see ^ 
Will need an allow string (which has ability to have quoted strings in it (iirc)) parser

File: packages/coil-extension/src/background/services/BackgroundFramesService.ts [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+          details.frameId,
+          partial
+        )
+        if (this.getFrame(details.tabId, details.frameId)?.state == null) {
+          this.requestFrameState(details.tabId, details.frameId)
+        }
+      }
+    }
+
+    this.api.webNavigation.onHistoryStateUpdated.addListener(
```
[the remark this answers is not in the document]
Comment (sublimator) [reply packages/coil-extension/src/background/services/BackgroundFramesService.ts]: Note that all the webNavigation apis are used from the *background* page. I didn't really see any perf issues, however we could certainly optimize. 

This service was an experiment to see if I could cobble something together that could keep track of a list of frames and their current url/state (state which we are currently using in our tabs.onUpdated handler )

File: packages/coil-extension/src/content/services/Frames.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (sublimator) [reply packages/coil-extension/src/content/services/Frames.ts]: e.g. 
`allow="vibrate https://a.com https://b.com; fullscreen 'none'; payment 'src'; payment; usb '*'"`

https://www.w3.org/TR/feature-policy-1/#parse-policy-directive

File: packages/coil-extension/test/fixtures/iframes/index.html [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  </head>
+  <body>
+    <iframe allow="monetization" id="frame1" src="./frame1.html"></iframe>
+    <iframe
+      allow="monetization another"
+      id="frame2"
+      src="./frame2.html"
+    ></iframe>
+    <iframe
+      allow="monetization"
```
Comment (sublimator) [inline packages/coil-extension/test/fixtures/iframes/index.html]: 2 allows hrmmmm

File: packages/coil-extension/src/background/services/BackgroundFramesService.ts [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+
+  updateOrAddFrame(
+    from: string,
+    tabId: number,
+    frameId: number,
+    partial: Readonly<Partial<Frame>>
+  ) {
+    const lastUpdateTimeMS = partial.lastUpdateTimeMS ?? Date.now()
+    const frame = this.getFrame(tabId, frameId)
+    if (frame && frame.lastUpdateTimeMS > lastUpdateTimeMS) {
```
Comment (sublimator) [inline packages/coil-extension/src/background/services/BackgroundFramesService.ts]: seems pointless

File: packages/coil-extension/src/background/services/BackgroundFramesService.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (sentientwaffle) [reply packages/coil-extension/src/background/services/BackgroundFramesService.ts]: Performance seems ok, though I agree we should be wary of this. Maybe `makeCallback` could early exit if the tab is not monetized?

**Comment by sublimator** [comment]
@sentientwaffle

re: optimization

Do note that this is just a first attempt at cobbling something together
that can track frames and their status and the PR is in draft status. I
wasn't sure this was possible until disovered and explored the
webNavigation api.

I did however wire this up to replace our (problematic) tabs.onUpdated
usage. As you saw there's really not much noticeable perf degration even
tracking all the frames.

I was planning on making any optimizations once the frames are actually
monetizing and the the supporting data structures have written themselves.

re: this.log field override

It's just "in development" hackery. Do note that visionmedia debug module
(with localStorage.debug set filtering ) doesn't give you proper source
file:line locations.

**Comment by sublimator** [comment]
Oh, re: %s string format usage, it doesn't work with the
console.log.bind(console, namespace) hack which supports source locations,
so tend to shy away from them.

Do you know any good debug library that works in the browser with proper
source locations? (Even if it uses some Error().stack hack for development)

(p.s. sorry for not replying in place. On mobile currently)

**Comment by sentientwaffle** [comment]
I don't know of one, sorry. I usually just use distinguishable log messages instead.

**Comment by sublimator** [comment]
Any thoughts on approach for impl outlined in initial comment ?

File: packages/coil-extension/src/content/content.ts [hunk 6]
```diff
@@ -38,6 +39,12 @@ function main() {
     configureContainer(container)
     container.get(ContentScript).init()
   }
+  console.log(
```
Comment (sublimator) [inline packages/coil-extension/src/content/content.ts]: Remove this as per #342

File: packages/coil-extension/src/content/services/ContentScript.ts [hunk 7]
```diff
@@ -156,7 +168,36 @@ export class ContentScript {
...
+        const data = event.data
+        if (typeof data.wmIframe === 'object') {
+          const {
+            forId,
+            allowed
+          }: { forId: string; allowed: boolean } = data.wmIframe
+          if (forId === this.uuid) {
+            if (allowed && this.monetization.hasRequest()) {
+              this.runtime.sendMessage(
+                this.monetization.startWebMonetizationMessage(forId)
```
Comment (sublimator) [inline packages/coil-extension/src/content/services/ContentScript.ts]: startWebMonetization message doesn't *need* the id really

**Pull request closed without merging**